core-plugins/src/main/java/io/cdap/plugin/batch/source/FTPBatchSource.java

```diff
@@ -130,6 +138,29 @@ public String getReferenceName() {
 
     @Override
     public String getPath() {
+      if (authContainsSpecialCharacters()) {
+        Path urlInfo;
+        String extractedPassword = extractPasswordFromUrl();
+        String encodedPassword = URLEncoder.encode(extractedPassword);
+        String validatePath = path.replace(extractedPassword, encodedPassword);
+        try {
+          urlInfo = new Path(validatePath);
+        } catch (Exception e) {
+          throw new IllegalArgumentException(String.format("Unable to parse url: %s %s", e.getMessage(), e));
+        }
+        String host = urlInfo.toUri().getAuthority().substring(urlInfo.toUri().getAuthority().lastIndexOf("@") + 1);
+        String user = urlInfo.toUri().getAuthority().split(":")[0];
+        String protocol = urlInfo.toUri().getScheme();
+        int port = urlInfo.toUri().getPort();
+        if (port == -1 && protocol.equals(FTP_PROTOCOL)) {
+          port = DEFAULT_FTP_PORT;
+        }
+        if (port == -1 && protocol.equals(SFTP_PROTOCOL)) {
+          port = DEFAULT_SFTP_PORT;
+        }
+        String cleanHost = host.replace(":" + port, "");
```

Comment:
We are removing the port from path here because of the format that is required in FTP File System when you add the necessary parameters as _File System Properties_. 

For example this will fail in File System Properties if we don't remove the port because it doesn't match the required format; 

`ftp://kimi:47@79]gfuss@192.168.0.179:21/kimi-look-here.txt`

"fs.ftp.host" --> "fs.ftp.host.192.168.0.179:21" 
"fs.ftp.user." + host  --> "fs.ftp.user.192.168.0.179:21.kimi" 
"fs.ftp.password." + host --> "fs.ftp.password.192.168.0.179:21.47@79]gfuss" 

The FTP or SFTP File system contains a **"fs.ftp.host.port"** which defaults to 21 in _FTP_ or 22 in _SFTP_ if the user doesn't specify the port in neither url or in File system properties, but if apparently we don't remove it from `host`, the job will fail. 

I've tested this specific scenario in: 
- CDAP 6.3.0 
- CDAP 6.1.4 
- CDAP 6.1.2 
- CDAP 6.2.0 

...and removing the port from host and adding it as standalone **"fs.ftp.host.port"** in File System Properties was the only way to get the job through.

Reply:
I see, thanks for the explanation!